File: 프로그래머스/3/118668. 코딩 테스트 공부/코딩 테스트 공부.js
function solution(alp, cop, problems) {
  let maxAlp = 0, maxCop = 0;
  for (const [aReq, cReq] of problems) {
    if (aReq > maxAlp) maxAlp = aReq;
    if (cReq > maxCop) maxCop = cReq;
  }

  alp = Math.min(alp, maxAlp);
  cop = Math.min(cop, maxCop);

  const INF = 1e9;
  const dp = Array.from({ length: maxAlp + 1 }, () => Array(maxCop + 1).fill(INF));
  dp[alp][cop] = 0;

  for (let a = alp; a <= maxAlp; a++) {
    for (let c = cop; c <= maxCop; c++) {
      const cur = dp[a][c];
      if (cur === INF) continue;

      if (a + 1 <= maxAlp) dp[a + 1][c] = Math.min(dp[a + 1][c], cur + 1);
      if (c + 1 <= maxCop) dp[a][c + 1] = Math.min(dp[a][c + 1], cur + 1);

      for (const [reqA, reqC, rwdA, rwdC, cost] of problems) {
        if (a >= reqA && c >= reqC) {
          const na = Math.min(maxAlp, a + rwdA);
          const nc = Math.min(maxCop, c + rwdC);
          dp[na][nc] = Math.min(dp[na][nc], cur + cost);
        }
      }
    }
  }

  return dp[maxAlp][maxCop];
}

File: 프로그래머스/3/118668. 코딩 테스트 공부/코딩 테스트 공부.java
import java.util.*;

class Solution {
    public int solution(int alp, int cop, int[][] problems) {
        int maxAlp = 0, maxCop = 0;
        for (int[] p : problems) {
            maxAlp = Math.max(maxAlp, p[0]);
            maxCop = Math.max(maxCop, p[1]);
        }

        alp = Math.min(alp, maxAlp);
        cop = Math.min(cop, maxCop);

        final int INF = 1_000_000_000;
        int[][] dp = new int[maxAlp + 1][maxCop + 1];
        for (int i = 0; i <= maxAlp; i++) Arrays.fill(dp[i], INF);
        dp[alp][cop] = 0;

        for (int a = alp; a <= maxAlp; a++) {
            for (int c = cop; c <= maxCop; c++) {
                int cur = dp[a][c];
                if (cur == INF) continue;

                if (a + 1 <= maxAlp) dp[a + 1][c] = Math.min(dp[a + 1][c], cur + 1);
                if (c + 1 <= maxCop) dp[a][c + 1] = Math.min(dp[a][c + 1], cur + 1);

                for (int[] p : problems) {
                    int reqA = p[0], reqC = p[1], rwdA = p[2], rwdC = p[3], cost = p[4];
                    if (a >= reqA && c >= reqC) {
                        int na = Math.min(maxAlp, a + rwdA);
                        int nc = Math.min(maxCop, c + rwdC);
                        dp[na][nc] = Math.min(dp[na][nc], cur + cost);
                    }
                }
            }
        }
        return dp[maxAlp][maxCop];
    }
}